File: betting/survivor.py
"""Survivor / Eliminator pool planner.

A survivor pool: each week you pick ONE team to win its game; if it loses you're
out; and you may use each team AT MOST ONCE all season.  Picking this week's
biggest favorite is a trap -- you'll want that team for a week when your options
are thin.  The right question is which *assignment of teams to weeks* keeps you
alive the longest.

We reuse the betting ``Projector`` for a market-anchored win probability on every
remaining game, then solve the season as a max-weight bipartite assignment
(weeks x teams, each team used once) that maximizes total survival probability
-- equivalently, sum of log(win_prob).  That plan tells you what to pick *now*
while reserving strong teams for the weeks you'll need them.  Recompute weekly as
results and lines move.

Free data only: win probabilities come from the same ESPN-odds-anchored model as
the Betting tab; no paid feeds.
"""

from __future__ import annotations

import math

from betting import data
from betting.model import Projector

LAST_REG_WEEK = 18
_NEG = 1e9  # cost for an impossible (team-not-playing) cell


# ---------------------------------------------------------------------------
# Rectangular Hungarian algorithm (Kuhn-Munkres), minimizes total cost.
# rows <= cols.  Returns row -> col assignment (0-indexed).  O(rows^2 * cols).
# ---------------------------------------------------------------------------
def _hungarian(cost: list[list[float]]) -> list[int]:
    n = len(cost)
    if n == 0:
        return []
    m = len(cost[0])
    INF = float("inf")
    u = [0.0] * (n + 1)
    v = [0.0] * (m + 1)
    p = [0] * (m + 1)          # p[j] = row assigned to column j (1-indexed)
    way = [0] * (m + 1)
    for i in range(1, n + 1):
        p[0] = i
        j0 = 0
        minv = [INF] * (m + 1)
        used = [False] * (m + 1)
        while True:
            used[j0] = True
            i0 = p[j0]
            delta = INF
            j1 = -1
            for j in range(1, m + 1):
                if not used[j]:
                    cur = cost[i0 - 1][j - 1] - u[i0] - v[j]
                    if cur < minv[j]:
                        minv[j] = cur
                        way[j] = j0
                    if minv[j] < delta:
                        delta = minv[j]
                        j1 = j
            for j in range(0, m + 1):
                if used[j]:
                    u[p[j]] += delta
                    v[j] -= delta
                else:
                    minv[j] -= delta
            j0 = j1
            if p[j0] == 0:
                break
        while j0:
            j1 = way[j0]
            p[j0] = p[j1]
            j0 = j1
    ans = [-1] * n
    for j in range(1, m + 1):
        if p[j] > 0:
            ans[p[j] - 1] = j - 1
    return ans


def _optimal(weeks: list[int], week_prob: dict, teams: list[str]) -> tuple[float, dict]:
    """Best assignment of one distinct team per week. Returns (total_log_prob,
    {week: team}).  `week_prob[w]` is {team_abbr: win_prob} for that week."""
    if not weeks:
        return 0.0, {}
    cost = []
    for w in weeks:
        tp = week_prob[w]
        row = []
        for t in teams:
            wp = tp.get(t)
            row.append(-math.log(max(wp, 1e-6)) if wp is not None else _NEG)
        cost.append(row)
    assign = _hungarian(cost)
    plan, total = {}, 0.0
    for i, w in enumerate(weeks):
        col = assign[i]
        t = teams[col] if 0 <= col < len(teams) else None
        wp = week_prob[w].get(t) if t else None
        if wp is None:                       # infeasible cell -> best free team
            free = {x: week_prob[w][x] for x in week_prob[w] if x in teams and x not in plan.values()}
            if free:
                t = max(free, key=free.get)
                wp = free[t]
        plan[w] = t
        if wp:
            total += math.log(max(wp, 1e-6))
    return total, plan


# ---------------------------------------------------------------------------
# Board build
# ---------------------------------------------------------------------------
def _collect_weeks(proj: Projector):
    """Project every remaining regular-season week once. Returns
    (week_prob, week_games, teams_meta, detected_start)."""
    week_prob: dict = {}
    week_games: dict = {}
    teams_meta: dict = {}
    detected_start = None
    for wk in range(1, LAST_REG_WEEK + 1):
        b = data.scoreboard("nfl", seasontype=2, week=wk)
        tp: dict = {}
        games = []
        any_upcoming = False
        for g in b["games"]:
            ha, aa = g["home"]["abbr"], g["away"]["abbr"]
            for side in ("home", "away"):
                t = g[side]
                teams_meta.setdefault(t["abbr"], {"abbr": t["abbr"], "name": t.get("name"),
                                                  "full": t.get("full"), "logo": t.get("logo")})
            if g["completed"]:
                continue
            any_upcoming = True
            pr = proj.project(g)
            hp, ap = pr["home_win_prob"], pr["away_win_prob"]
            tp[ha], tp[aa] = hp, ap
            games.append({
                "away": aa, "home": ha,
                "away_name": g["away"].get("name"), "home_name": g["home"].get("name"),
                "away_logo": g["away"].get("logo"), "home_logo": g["home"].get("logo"),
                "away_wp": round(ap, 4), "home_wp": round(hp, 4),
                "fav": ha if hp >= ap else aa, "fav_wp": round(max(hp, ap), 4),
                "has_odds": bool(g["espn_odds"]), "date": g["date"],
            })
        week_prob[wk] = tp
        week_games[wk] = games
        if any_upcoming and detected_start is None:
            detected_start = wk
    if detected_start is None:
        detected_start = 1
    return week_prob, week_games, teams_meta, detected_start


def build_survivor(used=None, start_week=None, horizon=None) -> dict:
    used = set((a or "").upper() for a in (used or []))
    proj = Projector("nfl")
    week_prob, week_games, teams_meta, detected_start = _collect_weeks(proj)

    start = int(start_week) if start_week else detected_start
    start = max(1, min(start, LAST_REG_WEEK))
    end = LAST_REG_WEEK
    if horizon:
        end = min(LAST_REG_WEEK, start + int(horizon) - 1)
    weeks = [w for w in range(start, end + 1) if week_prob.get(w)]

    # candidate teams = everything playing in the horizon, minus already-used
    teamset = sorted({t for w in weeks for t in week_prob[w] if t not in used})

    # optimal full-horizon plan
    total_logp, plan_map = _optimal(weeks, week_prob, teamset)
    plan = []
    cum = 1.0
    for w in weeks:
        t = plan_map.get(w)
        wp = week_prob[w].get(t) if t else None
        cum *= (wp or 1.0)
        opp = _opponent(week_games[w], t)
        plan.append({"week": w, "team": t, "win_prob": round(wp, 4) if wp else None,
                     "opp": opp, "cum_survival": round(cum, 4)})

    # this-week options, each scored by season-long survival if forced now
    this_week = weeks[0] if weeks else None
    options = []
    if this_week is not None:
        rest = weeks[1:]
        for t in week_prob[this_week]:
            if t in used or week_prob[this_week][t] is None:
                continue
            sub_teams = [x for x in teamset if x != t]
            sub_total, _ = _optimal(rest, week_prob, sub_teams)
            season_logp = math.log(max(week_prob[this_week][t], 1e-6)) + sub_total
            options.append({
                "team": t, "win_prob": round(week_prob[this_week][t], 4),
                "opp": _opponent(week_games[this_week], t),
                "season_survival": math.exp(season_logp),
            })
        options.sort(key=lambda o: o["season_survival"], reverse=True)
        best = options[0]["season_survival"] if options else 0.0
        for o in options:
            o["cost_vs_best"] = round(best - o["season_survival"], 4)
            o["season_survival"] = round(o["season_survival"], 4)

    return {
        "league": "nfl",
        "detected_week": detected_start,
        "start_week": start,
        "end_week": end,
        "weeks_planned": len(weeks),
        "used": sorted(used),
        "teams": [teams_meta[a] for a in sorted(teams_meta)],
        "weeks": [{"week": w, "games": week_games[w]} for w in weeks],
        "plan": plan,
        "plan_survival": round(math.exp(total_logp), 4) if weeks else None,
        "recommended": options[0] if options else None,
        "this_week": {"week": this_week, "options": options[:14]},
        "model": {"teams_rated": len(proj.elo.teams),
                  "n_games_learned": proj.elo.meta.get("n_games", 0)},
    }


def _opponent(games: list, team: str | None) -> str | None:
    if not team:
        return None
    for g in games:
        if g["home"] == team:
            return "vs " + g["away"]
        if g["away"] == team:
            return "@ " + g["home"]
    return None

File: web/survivor.js
/* Survivor / Eliminator pool planner -- self-contained module.
 *
 * Each week pick one team to win; a team can be used only once all season.
 * Talks to POST /api/survivor, which returns market-anchored win probabilities
 * for every remaining game and the season-long optimal team-to-week assignment
 * (so you don't burn a strong team you'll need later). Hooks into the app's tab
 * switcher via window.Survivor.show().
 */
(function () {
  "use strict";

  var state = {
    board: null,
    loading: false,
    error: null,
    picks: [],        // [{week, team}] locked picks, in order
    extraUsed: [],    // teams marked used with no specific week
    horizon: 0,       // 0 = rest of season
    showWeeks: false,
    showGrid: false,
  };

  var LS_PICKS = "ff_survivor_picks";
  var LS_EXTRA = "ff_survivor_extra";

  function panel() { return document.getElementById("survivor-panel"); }
  function esc(s) { return (s == null ? "" : String(s)).replace(/[&<>"]/g, function (c) { return { "&": "&amp;", "<": "&lt;", ">": "&gt;", '"': "&quot;" }[c]; }); }
  function pct(x) { return x == null ? "—" : (100 * x).toFixed(1) + "%"; }

  function load() {
    try { state.picks = JSON.parse(localStorage.getItem(LS_PICKS)) || []; } catch (e) { state.picks = []; }
    try { state.extraUsed = JSON.parse(localStorage.getItem(LS_EXTRA)) || []; } catch (e) { state.extraUsed = []; }
  }
  function savePicks() { localStorage.setItem(LS_PICKS, JSON.stringify(state.picks)); }
  function saveExtra() { localStorage.setItem(LS_EXTRA, JSON.stringify(state.extraUsed)); }

  function usedTeams() {
    var s = {};
    state.picks.forEach(function (p) { s[p.team] = 1; });
    state.extraUsed.forEach(function (t) { s[t] = 1; });
    return Object.keys(s);
  }
  function startWeek() {
    if (!state.picks.length) return null; // auto-detect on the server
    var mx = 0; state.picks.forEach(function (p) { if (p.week > mx) mx = p.week; });
    return mx + 1;
  }

  function api(path, body) {
    var ctrl = new AbortController();
    var to = setTimeout(function () { ctrl.abort(); }, 90000);
    return fetch(path, { method: "POST", headers: { "Content-Type": "application/json" }, body: JSON.stringify(body || {}), signal: ctrl.signal })
      .then(function (r) { clearTimeout(to); if (!r.ok) throw new Error("HTTP " + r.status); return r.json(); },
        function (e) { clearTimeout(to); throw (e && e.name === "AbortError") ? new Error("timed out (first load fetches the full schedule; try again)") : e; });
  }

  function reload() {
    state.loading = true; state.error = null; render();
    api("/api/survivor", { used: usedTeams(), start_week: startWeek(), horizon: state.horizon || null })
      .then(function (b) { state.board = b; state.loading = false; render(); })
      .catch(function (e) { state.loading = false; state.error = e && e.message || String(e); render(); });
  }

  var initialized = false;
  function show() {
    if (!initialized) { load(); initialized = true; reload(); return; }
    render();
    if (!state.board && !state.loading) reload();
  }

  // ---- actions -------------------------------------------------------
  function lockPick(week, team) {
    state.picks = state.picks.filter(function (p) { return p.week !== week; });
    state.picks.push({ week: week, team: team });
    state.picks.sort(function (a, b) { return a.week - b.week; });
    savePicks();
    reload();
  }
  function undoPick(week) {
    state.picks = state.picks.filter(function (p) { return p.week !== week; });
    savePicks();
    reload();
  }
  function toggleExtra(team) {
    var i = state.extraUsed.indexOf(team);
    if (i >= 0) state.extraUsed.splice(i, 1); else state.extraUsed.push(team);
    saveExtra();
    reload();
  }
  function resetAll() {
    if (!confirm("Clear all your survivor picks and used teams?")) return;
    state.picks = []; state.extraUsed = []; savePicks(); saveExtra(); reload();
  }

  // ---- render --------------------------------------------------------
  function render() {
    var p = panel(); if (!p) return;
    if (state.error) {
      p.innerHTML = head() +
        '<div class="bet-err"><p><strong>Couldn’t build the survivor plan.</strong> ' + esc(state.error) + "</p>" +
        '<p class="muted">If the server was just updated, stop and relaunch <code>run.bat</code>, then reload. The first build fetches all 18 weeks (~20s) and is cached after.</p>' +
        '<button id="sv-retry" class="btn">↻ Retry</button></div>';
      var rb = p.querySelector("#sv-retry"); if (rb) rb.onclick = reload;
      return;
    }
    if (state.loading || !state.board) {
      p.innerHTML = head() + '<p class="muted">⏳ Projecting every remaining game and solving the season-long plan…</p>';
      wire();
      return;
    }
    var b = state.board;
    p.innerHTML = head() + recBanner(b) + pickLedger(b) + optionsTable(b) + planTable(b) + weeksSection(b) + gridSection(b) + method();
    wire();
  }

  function head() {
    return '<div class="sv-head">' +
      '<h2>🏆 Survivor Pool Planner</h2>' +
      '<p class="muted">Pick one team to win each week — but each team only once all season. This plans the whole season at once so you don’t burn a strong team you’ll need later.</p>' +
      '</div>';
  }

  function recBanner(b) {
    var r = b.recommended;
    if (!r) return '<div class="rec-banner">All planned weeks have a pick recorded. 🎉</div>';
    var alt = (b.this_week.options || [])[1];
    var altTxt = alt ? ' Nearly as good: <strong>' + esc(alt.team) + '</strong> ' + esc(alt.opp) + ' (' + pct(alt.win_prob) + ').' : "";
    return '<div class="rec-banner sv-rec">' +
      '🎯 <strong>Week ' + b.this_week.week + ' pick: ' + esc(r.team) + '</strong> ' + esc(r.opp) +
      ' — <strong>' + pct(r.win_prob) + '</strong> to win.' +
      '<div class="muted" style="margin-top:.3rem;font-size:.82rem">Chosen to maximize your chance of surviving the whole season, not just this week.' + altTxt + '</div>' +
      '</div>';
  }

  function pickLedger(b) {
    if (!state.picks.length) return "";
    var rows = state.picks.slice().sort(function (a, c) { return a.week - c.week; }).map(function (p) {
      return '<span class="sv-pick-chip">Wk ' + p.week + ': <strong>' + esc(p.team) + '</strong> <button class="sv-x" data-undo="' + p.week + '" title="undo">✕</button></span>';
    }).join("");
    return '<div class="sv-ledger"><span class="muted">Your locked picks:</span> ' + rows + '</div>';
  }

  function optionsTable(b) {
    var opts = (b.this_week && b.this_week.options) || [];
    if (!opts.length) return "";
    var rows = opts.map(function (o, i) {
      var rec = i === 0;
      return '<tr class="' + (rec ? "sv-best" : "") + '">' +
        '<td class="left"><span class="pos-badge pos-DEF">' + esc(o.team) + '</span> <span class="muted">' + esc(o.opp) + '</span>' + (rec ? ' <span class="sv-tag">best</span>' : "") + '</td>' +
        '<td><strong>' + pct(o.win_prob) + '</strong></td>' +
        '<td>' + pct(o.season_survival) + '</td>' +
        '<td class="' + (o.cost_vs_best > 0.0005 ? "val-neg" : "val-pos") + '">' + (o.cost_vs_best > 0.0005 ? "−" + (100 * o.cost_vs_best).toFixed(2) + "%" : "—") + '</td>' +
        '<td><button class="btn sv-lock" data-lock-team="' + esc(o.team) + '" data-lock-week="' + b.this_week.week + '">✓ Lock</button></td>' +
        '</tr>';
    }).join("");
    return '<div class="section-h">Week ' + b.this_week.week + ' — your best options</div>' +
      '<p class="muted" style="font-size:.78rem">“Season survival” = your chance of surviving the whole rest of the year if you pick this team now and follow the optimal plan afterward. “Cost” = how much that season-survival drops vs. the best option.</p>' +
      '<div class="table-wrap"><table class="lineup"><thead><tr><th class="left">Team</th><th>Win % this wk</th><th>Season survival</th><th>Cost</th><th>Pick</th></tr></thead><tbody>' + rows + '</tbody></table></div>';
  }

  function planTable(b) {
    var rows = (b.plan || []).map(function (pk) {
      var picked = state.picks.some(function (x) { return x.week === pk.week && x.team === pk.team; });
      return '<tr>' +
        '<td>Wk ' + pk.week + (pk.week === b.this_week.week ? ' <span class="sv-tag">now</span>' : "") + '</td>' +
        '<td class="left"><span class="pos-badge pos-DEF">' + esc(pk.team || "—") + '</span> <span class="muted">' + esc(pk.opp || "") + '</span>' + (picked ? ' ✓' : "") + '</td>' +
        '<td>' + pct(pk.win_prob) + '</td>' +
        '<td class="muted">' + pct(pk.cum_survival) + '</td>' +
        '</tr>';
    }).join("");
    var surv = b.plan_survival != null ? '<span class="muted">Full-season survival if you follow this plan: <strong>' + pct(b.plan_survival) + '</strong> (surviving all ' + b.weeks_planned + ' weeks is rare — most pools are won earlier).</span>' : "";
    return '<div class="section-h">Optimal season plan <span class="muted" style="font-size:.72rem;text-transform:none">· weeks ' + b.start_week + '–' + b.end_week + '</span></div>' +
      '<div class="table-wrap"><table class="lineup"><thead><tr><th>Week</th><th class="left">Suggested team</th><th>Win %</th><th>Cumulative survival</th></tr></thead><tbody>' + rows + '</tbody></table></div>' +
      '<p style="margin-top:.4rem">' + surv + '</p>';
  }

  function weeksSection(b) {
    var btn = '<button id="sv-toggle-weeks" class="btn">' + (state.showWeeks ? "▾ Hide" : "▸ Show") + ' every game’s win probability</button>';
    if (!state.showWeeks) return '<div class="section-h">Weekly slates</div>' + btn;
    var blocks = (b.weeks || []).map(function (w) {
      var games = w.games.slice().sort(function (a, c) { return c.fav_wp - a.fav_wp; }).map(function (g) {
        var noOdds = g.has_odds ? "" : ' <span class="muted" title="model-only, no market line yet">~</span>';
        return '<div class="sv-game"><span><span class="pos-badge pos-DEF">' + esc(g.fav) + '</span> ' + pct(g.fav_wp) + noOdds + '</span>' +
          '<span class="muted">' + esc(g.away) + " @ " + esc(g.home) + '</span></div>';
      }).join("");
      return '<div class="sv-week"><div class="sv-week-h">Week ' + w.week + '</div>' + games + '</div>';
    }).join("");
    return '<div class="section-h">Weekly slates <span class="muted" style="font-size:.72rem;text-transform:none">· favorite & win prob; ~ = model-only (no line yet)</span></div>' + btn +
      '<div class="sv-weeks">' + blocks + '</div>';
  }

  function gridSection(b) {
    var btn = '<button id="sv-toggle-grid" class="btn">' + (state.showGrid ? "▾ Hide" : "▸ Show") + ' “teams already used” editor</button>';
    if (!state.showGrid) return '<div class="section-h">Already used teams</div><p class="muted" style="font-size:.8rem">Locking picks above marks teams used automatically. Use this only to mark teams you burned before you started using this planner.</p>' + btn;
    var pickedTeams = {}; state.picks.forEach(function (p) { pickedTeams[p.team] = p.week; });
    var chips = (b.teams || []).map(function (t) {
      var inPick = pickedTeams[t.abbr];
      var extra = state.extraUsed.indexOf(t.abbr) >= 0;
      var cls = inPick ? "sv-chip used locked" : (extra ? "sv-chip used" : "sv-chip");
      var title = inPick ? "locked as your Week " + inPick + " pick" : (extra ? "marked used" : "available");
      return '<button class="' + cls + '" data-extra="' + esc(t.abbr) + '" ' + (inPick ? "disabled" : "") + ' title="' + esc(title) + '">' + esc(t.abbr) + (inPick ? " ✓" : "") + '</button>';
    }).join("");
    return '<div class="section-h">Already used teams</div>' + btn + '<div class="sv-grid">' + chips + '</div>';
  }

  function method() {
    var b = state.board || {};
    var m = b.model || {};
    return '<div class="method" style="margin-top:1rem"><strong>How this works:</strong> every remaining game gets a win probability from the same model as the Betting tab — team ratings (Elo + opponent-adjusted scoring) anchored to the live sportsbook line where one exists. The season is then solved as an assignment problem: one distinct team per week, maximizing the product of weekly win probabilities (your survival odds). That’s why the Week-' + (b.this_week ? b.this_week.week : "1") + ' pick isn’t always the biggest favorite — a big favorite is often worth saving for a week when your other options are weak. Recompute weekly as lines and results move. <span class="muted">Model: ' + (m.teams_rated || 0) + ' teams, ' + (m.n_games_learned || 0) + ' games learned. Free data only (ESPN).</span></div>';
  }

  function wire() {
    var p = panel(); if (!p) return;
    var h = document.getElementById("sv-horizon");
    if (h) h.onchange = function () { state.horizon = parseInt(h.value, 10) || 0; reload(); };
    var reset = document.getElementById("sv-reset");
    if (reset) reset.onclick = resetAll;
    var tw = document.getElementById("sv-toggle-weeks");
    if (tw) tw.onclick = function () { state.showWeeks = !state.showWeeks; render(); };
    var tg = document.getElementById("sv-toggle-grid");
    if (tg) tg.onclick = function () { state.showGrid = !state.showGrid; render(); };
    p.querySelectorAll(".sv-lock").forEach(function (btn) {
      btn.onclick = function () { lockPick(parseInt(btn.dataset.lockWeek, 10), btn.dataset.lockTeam); };
    });
    p.querySelectorAll("[data-undo]").forEach(function (btn) {
      btn.onclick = function () { undoPick(parseInt(btn.dataset.undo, 10)); };
    });
    p.querySelectorAll("[data-extra]").forEach(function (btn) {
      btn.onclick = function () { if (!btn.disabled) toggleExtra(btn.dataset.extra); };
    });
  }

  // The horizon control lives in the header so it's always present.
  var _head = head;
  head = function () {
    var opts = [[0, "Rest of season"], [6, "Next 6 weeks"], [4, "Next 4 weeks"], [3, "Next 3 weeks"]]
      .map(function (o) { return '<option value="' + o[0] + '"' + (state.horizon === o[0] ? " selected" : "") + '>' + o[1] + "</option>"; }).join("");
    return _head() +
      '<div class="league-bar" style="margin-bottom:.6rem">' +
      '<label class="muted">Plan horizon: <select id="sv-horizon">' + opts + '</select></label>' +
      '<button id="sv-reset" class="btn">Reset picks</button>' +
      '</div>';
  };

  window.Survivor = { show: show };
})();
